File: days_bw_dates.py
import datetime as dt
import calendar

_leap_months = {1:31, 2:29, 3:31, 4:30, 5:31, 6:30, 7:31, 8:31,
                9:30, 10:31, 11:30, 12:31}
_months = {1:31, 2:28, 3:31, 4:30, 5:31, 6:30, 7:31, 8:31,
           9:30, 10:31, 11:30, 12:31}

d1, m1, y1 = map(int, input().split(' '))
d2, m2, y2 = map(int, input().split(' '))

# initial, final = dt.date(y1, m1, d1), dt.date(y2, m2, d2)

# print("days :", (final-initial).days)

def is_leap_year(year):
    return calendar.isleap(year)

def days_in_month(month, is_leap):
    if is_leap:
        return _leap_months[month]
    else:
        return _months[month]

def total_months(m1, y1, m2, y2):
    """excluding m1 and m2
    """
    years = y2 - y1
    first = 12 - m1
    last = 12 - m2
    rest = years * 12
    return [m1+1, first - last + rest - 1]

def days_bw_dates(d1=14, m1=1, y1=2016, d2=28, m2=4, y2=2017): # delta 470
    total = 0

    # calculate days in first month
    initial_month_days = days_in_month(m1, is_leap_year(y1))
    initial_month_days = initial_month_days - d1
    if (y1 == y2 and m1 == m2):
        return d2 - d1

    # calculate days in rest_of_the_months
    month, no_months = total_months(m1, y1, m2, y2)
    for i in range(1, no_months+1):
        if month>12:
            month = month-12
            y1+=1
        total += days_in_month(month, is_leap_year(y1))
        month +=1

    # calculate days in last month ie d2
    return initial_month_days + total + d2

if __name__ == "__main__":
    #print(days_bw_dates(5, 5, 2017, 14, 6, 2017))
    #print(days_bw_dates(23, 4, 2017, 24, 8, 2017))
    print("no of days: ", days_bw_dates(d1, m1, y1, d2, m2, y2))
    #print(days_bw_dates())

File: days_bw_dates.c
#include<stdio.h>
#include<time.h>

int _leap_months[12] = {31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
int _months[12] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};

int *total_months(int m1, int y1, int m2, int y2) {
  int years, first, last, rest;
  static int result[2];

  years = y2 - y1;
  first = 12 - m1;
  last = 12 - m2;
  rest = years * 12;
  result[0] = m1 + 1;
  result[1] = first - last + rest - 1;

  return result;
}

int isleap(int year) {
  if(year%400 == 0)
    return 1;
  else if(year%100 == 0)
    return 0;
  else if(year%4 == 0)
    return 1;
  else
    return 0;
}


int days_in_month(int month, int year){
  int val = 0;

  if(month == 2)
    val = isleap(year);
  if(val == 1){
    return _leap_months[month-1];
  } else {
    return _months[month-1];
  }
}

int days_bw_dates(int d1, int m1, int y1, int d2, int m2, int y2) {
  int total, count, initial_month_days, month, no_months;
  int *p;

  total = 0;
  count = 0;
  initial_month_days = days_in_month(m1, y1);
  initial_month_days = initial_month_days - d1;
  if (y1 == y2 && m1 == m2){
    return d2 - d1;
  }
  p = total_months(m1, y1, m2, y2);
  month = p[0];
  no_months = p[1];
  for(int i = 1; i<no_months+1;i++){
    count = count + 1;
    if(month>12){
      month = month - 12;
      y1 = y1 + 1;
    }
    total = total + days_in_month(month, y1);
    month = month+1;
  }
  return initial_month_days + total + d2;
}

int main(){
  double t1;
  clock_t tic, toc;
  tic = clock();

  printf("days: %d\n", days_bw_dates(5, 5, 2017, 14, 6, 2017));
  printf("days: %d\n", days_bw_dates(23, 4, 2017, 24, 8, 2017));
  printf("days: %d\n", days_bw_dates(26, 8, 1993, 11, 1, 2021));
  printf("days: %d\n", days_bw_dates(26, 8, 1993, 15, 6, 2017));
  printf("days: %d\n", days_bw_dates(14, 1, 2016, 28, 4, 2017));

  toc = clock();
  t1 = (double) (toc-tic)/CLOCKS_PER_SEC;

  printf("avg time: %f\n", (double)t1/5);

  return 0;
}
